clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -253,6 +284,65 @@ public void forceClose() {
         initiateClose();
     }
 
+    private void getPID(long remainingTime) {
+        // Send an InitPIDRequest, wait for the response, retrieve the PID from the response if success. Return it.
+        // Otherwise throw exception.
+        // Pick a random broker to send the initPIDRequest to.
+        metadata.requestUpdate();
+        long start = time.milliseconds();
+        Node node = client.leastLoadedNode(time.milliseconds());
+        while (!client.ready(node, time.milliseconds()) && 0 <= remainingTime) {
+            if (client.connectionFailed(node)) {
+                return;
+            }
+            client.poll(100, time.milliseconds());
+            remainingTime = remainingTime - (time.milliseconds() - start);
+        }
+        if (remainingTime <= 0) {
+            return;
+        }
+
+        final AtomicBoolean initPidRequestCompleted = new AtomicBoolean(false);
+
+        String nodeId = node.idString();
+        InitPidRequest.Builder builder = new InitPidRequest.Builder(null);
+        RequestCompletionHandler callback = new RequestCompletionHandler() {
+            public void onComplete(ClientResponse response) {
+                initPidRequestCompleted.set(true);
+                handleInitPidResponse(response);
+            }
+        };
+
+        ClientRequest request = client.newClientRequest(nodeId, builder, time.milliseconds(), true, callback);
+        client.send(request, time.milliseconds());
+        start = time.milliseconds();
+        boolean isDisconnected = false;
+        while (!initPidRequestCompleted.get() && !isDisconnected && 0 <= remainingTime) {
+            client.poll(retryBackoffMs, time.milliseconds());
+            isDisconnected = client.connectionFailed(node);
+            remainingTime = remainingTime - (time.milliseconds() - start);
+        }
+    }
+
+    private void maybeWaitForPid() {
+        while (!transactionState.hasPid()) {
```

Comment:
We have a nested "while" loop here and I'm wondering if it is better to flatten in into a single while loop. Take `AbstractCoordinator.ensureCoordinatorReady` as an example, which does the following:

```
while (not found pid and not timed out) {

  if( findANode) {
     send()
     blockWaitForResponse
  } else {
     // either throw an exception, or continue to the next loop to retry
  }

  // back of if needed
}